sorting/merge_sort.c

```diff
@@ -1,139 +1,93 @@
+    i = 0;
+    j = 0;
+    k = left;
+    while (i < n1 && j < n2)
     {
-        if (a[p1] <= a[p2])
+        if (L[i] <= R[j])
         {
-            b[c++] = a[p1];
-            p1++;
+            arr[k] = L[i];
+            i++;
         }
         else
         {
-            b[c++] = a[p2];
-            p2++;
+            arr[k] = R[j];
+            j++;
         }
+        k++;
     }
 
-    if (p2 == r + 1)
+    // Copy the remaining elements of L and R, if any
+    while (i < n1)
     {
-        while ((p1 < ((l + r) / 2) + 1))
-        {
-            b[c++] = a[p1];
-            p1++;
-        }
+        arr[k] = L[i];
+        i++;
+        k++;
     }
-    else
+    while (j < n2)
     {
-        while ((p2 < r + 1))
-        {
-            b[c++] = a[p2];
-            p2++;
-        }
+        arr[k] = R[j];
+        j++;
+        k++;
     }
-
-    for (c = l; c < r + 1; c++) a[c] = b[c];
-
-    free(b);
 }
 
 /** Merge sort algorithm implementation
- * @param a array to sort
- * @param n number of elements in the array
- * @param l index to sort from
- * @param r index to sort till
+ * @param arr the array
+ * @param left index to sort from
+ * @param right index to sort till
  */
-void merge_sort(int *a, int n, int l, int r)
+void merge_sort(int arr[], int left, int right)
 {
-    if (r - l == 1)
-    {
-        if (a[l] > a[r])
-            swap(&a[l], &a[r]);
-    }
-    else if (l != r)
+    if (left < right)
     {
-        merge_sort(a, n, l, (l + r) / 2);
-        merge_sort(a, n, ((l + r) / 2) + 1, r);
-        merge(a, l, r, n);
-    }
+        int mid = left + (right - left) / 2;
 
-    /* no change if l == r */
+        // Sort the left and right halves separately
+        merge_sort(arr, left, mid);
+        merge_sort(arr, mid + 1, right);
+
+        // Merge the two sorted halves
+        merge(arr, left, mid, right);
+    }
 }
-/** @} */
 
-/** Main function */
-int main(void)
+int main()
 {
-    int *a, n, i;
-    printf("Enter Array size: ");
+    int n;
+    printf("Enter the number of elements: ");
     scanf("%d", &n);
-    if (n <= 0) /* exit program if arraysize is not greater than 0 */
-    {
-        printf("Array size must be Greater than 0!\n");
-        return 1;
-    }
-    a = (int *)malloc(n * sizeof(int));
-    if (a == NULL) /* exit program if can't malloc memory */
-    {
-        printf("Can't Malloc! Please try again.");
-        return 1;
-    }
-    for (i = 0; i < n; i++)
-    {
-        printf("Enter number[%d]: ", i);
-        scanf("%d", &a[i]);
-    }
 
-    merge_sort(a, n, 0, n - 1);
-    printf("Sorted Array: ");
-    for (i = 0; i < n; i++)
-    {
-        printf("%d ", a[i]);
-    }
-    printf("\n");
+    int arr[n];
+    printf("Enter the elements: ");
+    for (int i = 0; i < n; i++) scanf("%d", &arr[i]);
+
+    printf("Original array: ");
+    for (int i = 0; i < n; i++) printf("%d ", arr[i]);
+
+    merge_sort(arr, 0, n - 1);
```

Comment:
Please make a separate function for all the testing!
I am not that well read on c but if there is a static keyword please use that as well 
```c
static void tests(){
// tests here 
}
```

Reply:
I think the tests function is not available in another sorting algorithm. so why do we use only for this file?

Reply:
I have added the test function.